addons/mail/models/mail_render_mixin.py

```diff
@@ -206,6 +206,47 @@ def _render_encapsulate(self, layout_xmlid, html, add_context=None, context_reco
             html = self.env['mail.render.mixin']._replace_local_links(html)
         return html
 
+    @api.model
+    def _generate_preview(self, html):
+        """ Generate a preview of html, to used for example in marketing emails.
+
+        We add the character `&zwnj;` (zero-width non-joiner) to fill the end of
+        the preview in order to not have the beginning of the mail at the end of
+        the preview (don't work with simple space as the content is trimmed).
+        https://litmus.com/blog/the-ultimate-guide-to-preview-text-support
+
+        :param html: html content for which we generate the preview
+        :return: html to use as preview
+        """
+        preview = ""
+        for element in re.finditer(r"\>([^<>]+)\<", html):
+            element = element.group(1).strip()
+            if element:
+                preview += " %s" % element
+            if len(preview) > 500:
+                break
+
+        html_preview = f"""
+            <div style="display:none;font-size:1px;height:0px;width:0px;opacity:0;">
+              {tools.html_escape(preview)} {'&zwnj;&nbsp;' * 500}
```

Comment:
Considering that the average chars displayed in preview text is below 200 chars, adding 500 * '&zwnj;&nbsp;' means 6KB of junk bytes in *every single marketing email we send*, no? Ouch.

Could we use a more reasonable target of 250 chars and **subtract** from those the bytes that are already in the real preview text, considering that we try to grab 500 chars above.

And perhaps we could use the actual unicode chars `"\u200c\u00a0"` rather than the HTML entities (5 UTF8 bytes instead of 12). And do we really need both ZWNJ and NBSP?

Reply:
@tde-banana-odoo Additionally, I wonder why we want to generate a "preview text" of the actual HTML content, which is exactly what mail clients will be doing anyway. We're just duplicating the content.

It feels like the preview text would be useful only if we can put specific content that is not part of the visible mail, as a form of teaser or summary. In that case there will be no extraction to do, we just need to allow users to edit some text that is at the top of the email and always made invisible in the sent email.

See also https://litmus.com/blog/the-ultimate-guide-to-preview-text-support which explains the whole idea and is mentioned in the docstring...

Reply:
@odony 
cc @tde-banana-odoo 

If we let the mail client generate the summary based on the mail body (without adding an invisible HTML element an the beginning), it will add some unwanted garbage (e.g. the alt attribute of the images tag), see https://www.awesomescreenshot.com/image/4597869/2db6fae1112b29048a1e8ef5209f8503

We can maybe replace the HTML code with the unicode character to have less bytes as you said but we will need to test on all mail clients if it clean the end of the summary (but indeed, we can optimize it by removing extra chars).

I think also that we should be able to customize the preview text (but the original task was considered as a fix and not as an improvement, so we asked me to remove it).